react/states_and_effects/how_to_deal_with_side_effects.md

```diff
@@ -0,0 +1,242 @@
+
+*   You do not need to use an effect if you are only calculating something based on the state during rendering. For a change in a component, due to a change in the props, you can simply calculate and set it during rendering.
+
+    ~~~jsx
+    import React, { useState } from "react";
+
+    export default function AdditionDisplay() {
+      const [number1, setNumber1] = useState(0);
+      const [number2, setNumber2] = useState(0);
+
+      // This is all unnecessary.
+
+      // const [sum, setSum] = useState(0);
+      // useEffect(() => {
+      //   setSum(number1 + number2);
+      // }, [number1, number2]);
+
+      const sum = number1 + number2;
+
+      return (
+        <p>{number1} + {number2} = {sum}</p>
+      );
+    }
+    ~~~
+
+*   You do not need effects for events. Code that runs when a component is **displayed** should be in effects, the rest should be in events.
+
+    ~~~jsx
+    import React, { useState } from "react";
+
+    export default function App() {
+      const [input, setInput] = useState("");
+
+      const handleInput = (e) => {
+        setInput(e.target.value);
+      }
+
+      // You should avoid direct manipulation when not necessary
+
+      // useEffect(() => {
+      //   document.getElementById("name").addEventListener("click", handleInput);
+      //   return () => {
+      //     document.getElementById("name").removeEventListener("click", handleInput);
+      //   }
+      // });
+      
+      return (
+        <>
+          {/* <input id="name" /> */}
+
+          <input onChange={handleInput} value={input} />
+          <p>{ input }</p>
+        </>
+      )
+    }
+    ~~~
+
+*   You do not need an effect to reset the state based on a condition most of the time. You have learned about keys in React. Just like using a key on a list's item, adding one to a component, based on the state on which it should be reset creates a unique version of that component for each change in the value of the state.
+
+*   If you are having issues with managing your state and want to use an effect to update the state of a parent or some other non-child component, consider <span id="lifting-the-state">**lifting the state**.</span> As we know, in React, state flows in one direction, generally down the DOM. So the parents know of the data before passing it to the children. If multiple children are required to make use of a single state, it should be moved up to the parent that has all of the components that need it, instead of using escape hatches like an effect.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+1.  [This lesson of the React docs](https://react.dev/learn/lifecycle-of-reactive-effects) talks about the life of a component, the different stages at which rendering takes place, and the role of `useEffect` in it.
+2.  [This article from Epic React](https://epicreact.dev/why-you-shouldnt-put-refs-in-a-dependency-array) goes into how dependency arrays are supposed to be used.
```

Comment:
Refs aren't covered in the curriculum at this point. I'm not sure if we would want to include this resource in the assignment section

Reply:
Hmm, I think it could go into Additional Resources then? @01zulfi

Reply:
That works

Reply:
It's included in the commit itself :D